### PR TITLE
FB8-63: Command to purge gtid sets for given uuids

### PR DESCRIPTION
**Pull request by dutow**
Summary:

Jira ticket: https://jira.percona.com/browse/FB8-63

Reference Patch: https://github.com/facebook/mysql-5.6/commit/eb8a43a
Reference Patch: https://github.com/facebook/mysql-5.6/commit/70d9a6d

A new command 'purge gtid' is provided to remove old gtid sets for given uuids.
The command is logged to the binlog so that slaves will also purge their gtid states
making all the replicas consistent with master's state. The command takes multiple
uuids in the following form and can only be executed by super user.

  purge gtid '551dab6a-d12b-11e5-a3ea-0002c9d31542,551dab6a-d12b-11e5-a3ea-0002c9d31543'

Purging gtid sets generated by current server is not allowed since that will break slave's
auto-positioning. This command succeeds even if a given uuid is not present in the gtid set.
If writing to binlog fails, the client will receive an error. In these cases the gtid will
get purged regardless of success of binlog write. Since this operation is idempotent, this
behavior is safe and the client running the command can retry.

### CODE REVIEW

File: mysql-test/suite/rpl_gtid/t/rpl_purge_gtid.test [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+echo Check 'purge gtid' query resets logged and lost gtid state.;
+purge gtid '551dab6a-d12b-11e5-a3ea-0002c9d31542 , 551dab6a-d12b-11e5-a3ea-0002c9d31543';
+insert into t1 values(1);
+insert into t1 values(1);
+
+replace_result $master_uuid master_uuid;
+select @@global.gtid_executed;
+
+sync_slave_with_master;
```
Comment (laurynas-biveinis) [inline mysql-test/suite/rpl_gtid/t/rpl_purge_gtid.test]:
```suggestion
--source include/sync_slave_sql_with_master.inc
```

File: sql/rpl_gtid_set.cc [hunk 2]
```diff
@@ -304,6 +304,22 @@ void Gtid_set::clear_set_and_sid_map() {
   DBUG_VOID_RETURN;
 }
 
+void Gtid_set::remove(const std::vector<rpl_sidno> &sidnos) {
+  if (sid_lock != NULL) sid_lock->assert_some_wrlock();
```
Comment (laurynas-biveinis) [inline sql/rpl_gtid_set.cc]:
```suggestion
  if (sid_lock != nullptr) sid_lock->assert_some_wrlock();
```

File: sql/sql_parse.cc [hunk 3]
```diff
@@ -3203,6 +3206,61 @@ int mysql_execute_command(THD *thd, bool first_level) {
...
+        if (sid.parse(token, len) != RETURN_STATUS_OK) {
+          my_error(ER_INVALID_UUID, MYF(0), token);
+          goto error;
+        }
+        global_sid_lock->rdlock();
+        rpl_sidno sidno;
+        sidno = global_sid_map->sid_to_sidno(sid);
+        if (sidno == server_sidno || (master_sidno && sidno == master_sidno)) {
+          global_sid_lock->unlock();
+          my_error(ER_CANNOT_PURGE_SERVER_UUID, MYF(0), NULL);
```
Comment (laurynas-biveinis) [inline sql/sql_parse.cc]:
```suggestion
          my_error(ER_CANNOT_PURGE_SERVER_UUID, MYF(0), nullptr);
```

File: sql/sql_parse.cc [hunk 4]
```diff
@@ -3203,6 +3206,61 @@ int mysql_execute_command(THD *thd, bool first_level) {
...
+            gtid_state->get_previous_gtids_logged()->contains_sidno(sidno))
+          sidnos.push_back(sidno);
+        else {
+          push_warning_printf(thd, Sql_condition::SL_NOTE,
+                              ER_UUID_NOT_IN_EXECUTED_GTID_SET,
+                              ER_THD(thd, ER_UUID_NOT_IN_EXECUTED_GTID_SET),
+                              token);
+        }
+        global_sid_lock->unlock();
+        token = strtok(NULL, ",");
```
Comment (laurynas-biveinis) [inline sql/sql_parse.cc]:
```suggestion
        token = strtok(nullptr, ",");
```

**Review by laurynas-biveinis — CHANGES_REQUESTED** [review]

File: sql/rpl_gtid.h [hunk 5]
```diff
@@ -1344,6 +1345,14 @@ class Gtid_set {
     existing allocated memory will be re-used.
   */
   void clear();
+
+  /**
+    Removes the sidnos (and their intervals) from the logged state.
+
+    @params sidnos List of rpl_sidno to remove.
+  */
+  void remove(const std::vector<rpl_sidno> &sidnos);
```
Comment (percona-ysorokin) [inline sql/rpl_gtid.h]: typedef/using for a custom container type.
```suggestion
  using rpl_sidno_container = std::vector<rpl_sidno>;
  void remove(const rpl_sidno_container &sidnos);
```

File: sql/rpl_gtid_set.cc [hunk 6]
```diff
@@ -304,6 +304,22 @@ void Gtid_set::clear_set_and_sid_map() {
   DBUG_VOID_RETURN;
 }
 
+void Gtid_set::remove(const std::vector<rpl_sidno> &sidnos) {
```
Comment (percona-ysorokin) [inline sql/rpl_gtid_set.cc]:
```suggestion
void Gtid_set::remove(const rpl_sidno_container &sidnos) {
```

File: sql/rpl_gtid_set.cc [hunk 7]
```diff
@@ -304,6 +304,22 @@ void Gtid_set::clear_set_and_sid_map() {
...
+  if (!max_sidno) return;
+  Free_intervals_lock lock(this);
+  lock.lock_if_not_locked();
+  for (auto sidno : sidnos) {
+    if (sidno >= 1 && sidno <= max_sidno) {
+      Interval_iterator ivit(this, sidno);
+      if (ivit.get()) ivit.remove(this);
+    }
+  }
+  lock.unlock_if_locked();
```
Comment (percona-ysorokin) [inline sql/rpl_gtid_set.cc]: Redundant: `unlock_if_locked` will be called automatically in `lock`'s destructor.

File: sql/rpl_gtid_set.cc [hunk 8]
```diff
@@ -304,6 +304,22 @@ void Gtid_set::clear_set_and_sid_map() {
   DBUG_VOID_RETURN;
 }
 
+void Gtid_set::remove(const std::vector<rpl_sidno> &sidnos) {
+  if (sid_lock != NULL) sid_lock->assert_some_wrlock();
+  cached_string_length = -1;
+  rpl_sidno max_sidno = get_max_sidno();
+  if (!max_sidno) return;
```
Comment (percona-ysorokin) [inline sql/rpl_gtid_set.cc]: avoid implicit conversions
```suggestion
  if (max_sidno == 0) return;
```

File: sql/rpl_slave.cc [hunk 9]
```diff
@@ -673,6 +673,25 @@ int stop_slave(THD *thd) {
   DBUG_RETURN(error);
 }
 
+void get_master_sidno(THD *thd, rpl_sidno &master_sidno) {
+  // We should not purge gtids received from current active master.
+  channel_map.wrlock();
+  LEX *lex = thd->lex;
```
Comment (percona-ysorokin) [inline sql/rpl_slave.cc]:
```suggestion
  const auto lex = thd->lex;
```

File: sql/rpl_slave.cc [hunk 10]
```diff
@@ -673,6 +673,25 @@ int stop_slave(THD *thd) {
...
+
+  auto active_mi = channel_map.get_mi(
```
Comment (percona-ysorokin) [inline sql/rpl_slave.cc]:
```suggestion
  const auto active_mi = channel_map.get_mi(
```

File: sql/sql_parse.cc [hunk 11]
```diff
@@ -3203,6 +3206,61 @@ int mysql_execute_command(THD *thd, bool first_level) {
       res = purge_master_logs_before_date(thd, purge_time);
       break;
     }
+    case SQLCOM_PURGE_UUID: {
+      if (check_global_access(thd, SUPER_ACL)) goto error;
+      rpl_sidno server_sidno = gtid_state->get_server_sidno();
+      rpl_sidno master_sidno = 0;
+      // We should not purge gtids received from current active master.
+      get_master_sidno(thd, master_sidno);
+      std::vector<rpl_sidno> sidnos;
```
Comment (percona-ysorokin) [inline sql/sql_parse.cc]:
```suggestion
      rpl_sidno_container sidnos;
```

File: sql/sql_parse.cc [hunk 12]
```diff
@@ -3203,6 +3206,61 @@ int mysql_execute_command(THD *thd, bool first_level) {
...
+        auto len = strlen(token);
+        while (token[len - 1] == ' ') len--;
+        if (sid.parse(token, len) != RETURN_STATUS_OK) {
+          my_error(ER_INVALID_UUID, MYF(0), token);
+          goto error;
+        }
+        global_sid_lock->rdlock();
+        rpl_sidno sidno;
+        sidno = global_sid_map->sid_to_sidno(sid);
+        if (sidno == server_sidno || (master_sidno && sidno == master_sidno)) {
```
Comment (percona-ysorokin) [inline sql/sql_parse.cc]:
```suggestion
        if (sidno == server_sidno || (master_sidno != 0 && sidno == master_sidno)) {
```

File: sql/sql_parse.cc [hunk 13]
```diff
@@ -3203,6 +3206,61 @@ int mysql_execute_command(THD *thd, bool first_level) {
...
+          global_sid_lock->unlock();
+          my_error(ER_CANNOT_PURGE_SERVER_UUID, MYF(0), NULL);
+          goto error;
+        }
+        if (sidno &&
```
Comment (percona-ysorokin) [inline sql/sql_parse.cc]:
```suggestion
        if (sidno != 0 &&
```

File: sql/sql_parse.cc [hunk 14]
```diff
@@ -3203,6 +3206,61 @@ int mysql_execute_command(THD *thd, bool first_level) {
...
+                              ER_UUID_NOT_IN_EXECUTED_GTID_SET,
+                              ER_THD(thd, ER_UUID_NOT_IN_EXECUTED_GTID_SET),
+                              token);
+        }
+        global_sid_lock->unlock();
+        token = strtok(NULL, ",");
+      }
+      mysql_mutex_lock(mysql_bin_log.get_log_lock());
+      global_sid_lock->wrlock();
+      const_cast<Gtid_set *>(gtid_state->get_previous_gtids_logged())
```
Comment (percona-ysorokin) [inline sql/sql_parse.cc]: looks like a hack - consider extending `Gtid_state`' interface with a new method

**Review by percona-ysorokin — CHANGES_REQUESTED** [review]

**Review by percona-ysorokin — APPROVED** [review]
LGTM

File: share/errmsg-utf8.txt [hunk 15]
```diff
@@ -18724,6 +18724,15 @@ ER_BINLOG_CANT_READ_INDEX
 ER_RPL_FAILED_IN_RLI_INIT_INFO
   eng "Failed in %s called from Relay_log_info::rli_init_info()."
 
+ER_INVALID_UUID
```
Comment (hermanlee) [inline share/errmsg-utf8.txt]: These error codes should be mapped to the 5.6 section. I think this replaces ER_PLACEHOLDER_50030, with ER_UUID_NOT_IN_EXECUTED_GTID_SET replacing ER_PLACEHOLDER_50032.

File: sql/sql_parse.cc [hunk 16]
```diff
@@ -3203,6 +3206,60 @@ int mysql_execute_command(THD *thd, bool first_level) {
...
+      if (check_global_access(thd, SUPER_ACL)) goto error;
+      rpl_sidno server_sidno = gtid_state->get_server_sidno();
+      rpl_sidno master_sidno = 0;
+      // We should not purge gtids received from current active master.
+      get_master_sidno(thd, master_sidno);
+      rpl_sidno_container sidnos;
+      char *token = strtok(lex->gtid_string, ",");
+      while (token) {
+        rpl_sid sid;
+        while (*token == ' ') token++;
```
Comment (hermanlee) [inline sql/sql_parse.cc]: I think we may want the equivalent of SKIP_WHITESPACE here? \t and others were acceptable forms of whitespace to this command in 5.6.

**Review by hermanlee — CHANGES_REQUESTED** [review]
I think rpl_gtid.rpl_purge_gtid and main.mysqld--help-notwin tests are failing.

**Comment by dutow** [comment]
Looks like rpl_purge_gtid is unstable: sometimes the master_uuid is before the custom uuids, sometimes it is after.

File: sql/sql_parse.cc [hunk 17]
```diff
@@ -3203,6 +3206,60 @@ int mysql_execute_command(THD *thd, bool first_level) {
...
+      rpl_sidno server_sidno = gtid_state->get_server_sidno();
+      rpl_sidno master_sidno = 0;
+      // We should not purge gtids received from current active master.
+      get_master_sidno(thd, master_sidno);
+      rpl_sidno_container sidnos;
+      char *token = strtok(lex->gtid_string, ",");
+      while (token) {
+        rpl_sid sid;
+        while (my_isspace(&my_charset_utf8_general_ci, *token)) token++;
+        auto len = strlen(token);
```
Comment (hermanlee) [inline sql/sql_parse.cc]: Malformed gtid_strings with double commas like "uuid1:1-3,,uuid2:3-5" could result in len = 0. I'll can fix this before applying the change.

File: sql/mysqld.cc [hunk 18]
```diff
@@ -3544,6 +3544,9 @@ SHOW_VAR com_status_vars[] = {
     {"purge_before_date",
      (char *)offsetof(System_status_var, com_stat[(uint)SQLCOM_PURGE_BEFORE]),
      SHOW_LONG_STATUS, SHOW_SCOPE_ALL},
+    {"purge_gtid",
+     (char *)offsetof(System_status_var, com_stat[(uint)SQLCOM_PURGE_UUID]),
+     SHOW_LONG_STATUS},
```
Comment (hermanlee) [inline sql/mysqld.cc]: Also missing SHOW_SCOPE_ALL. I'll fix this before pushing.

**Review by hermanlee — APPROVED** [review]

File: sql/rpl_gtid_set.cc [hunk 19]
```diff
@@ -304,6 +304,21 @@ void Gtid_set::clear_set_and_sid_map() {
   DBUG_VOID_RETURN;
 }
 
+void Gtid_set::remove(const rpl_sidno_container &sidnos) {
+  if (sid_lock != nullptr) sid_lock->assert_some_wrlock();
+  cached_string_length = -1;
```
Comment (lth) [inline sql/rpl_gtid_set.cc:309]: It would seem like `has_cached_string_length = false` should be the statement you need if you want to invalidate the string length?

**Review by lth — CHANGES_REQUESTED** [review]

File: sql/sql_parse.cc (hunk 17 above)
Comment (dutow) [reply sql/sql_parse.cc]: Fixed, and added a test for this.

File: sql/mysqld.cc (hunk 18 above)
Comment (dutow) [reply sql/mysqld.cc]: Fixed.

File: sql/rpl_gtid_set.cc (hunk 19 above)
Comment (dutow) [reply sql/rpl_gtid_set.cc:309]: Fixed.

File: sql/sql_parse.cc [hunk 20]
```diff
@@ -3203,6 +3206,60 @@ int mysql_execute_command(THD *thd, bool first_level) {
...
+        while (my_isspace(&my_charset_utf8_general_ci, token[len - 1])) len--;
```
Comment (hermanlee) [inline sql/sql_parse.cc]: I think you also need to check for len == 0 here before decrementing len.

**Review by hermanlee — CHANGES_REQUESTED** [review]

File: sql/sql_parse.cc (hunk 20 above)
Comment (dutow) [reply sql/sql_parse.cc]: Done. I've also added a few additional test cases, and fixed some other corner cases, such as an empty string.
